File: integrations/langgraph/src/namespace.ts
/**
 * Namespace templating and user-id resolution.
 *
 * LangGraph addresses long-term memory by *namespace* — an array of strings such as
 * `["memories", "alice"]` (`@langchain/langgraph-checkpoint`, `src/store/base.ts`, `BaseStore.put`).
 * Cortadel addresses it by *user id*, fixed at client construction: "All calls are scoped to the
 * userId given at construction." Those two models meet here.
 *
 * - {@link resolveNamespace} expands `{placeholder}` labels from a `LangGraphRunnableConfig`'s
 *   `configurable` object, so a single tool or node can serve every user.
 * - {@link namespaceUserId} builds the `namespace -> userId` callable that {@link CortadelStore}
 *   uses to decide *which* Cortadel client a given namespace belongs to.
 */
import { ambientConfig } from "./runtime.js";

/**
 * A namespace, or a template for one. A bare `string` is treated as a one-label namespace.
 */
export type NamespaceLike = string | readonly string[];

/**
 * Either a fixed Cortadel user id, or a callable mapping a namespace to one.
 */
export type UserIdResolver = string | ((namespace: string[]) => string);

/**
 * The namespace this package defaults to.
 *
 * Mirrors the shape LangGraph's own long-term-memory docs use, and pairs with
 * `namespaceUserId(-1)` so the trailing label *is* the Cortadel user id.
 */
export const DEFAULT_NAMESPACE: readonly string[] = ["memories", "{userId}"];

// Only a *whole* label may be a placeholder ("{userId}", never "user-{id}"). Keeping that
// restriction means a namespace label can never be half-literal, which would make the reverse
// mapping (namespace -> user id) ambiguous.
const PLACEHOLDER = /^\{(\w+)\}$/;

/** Normalise a namespace-ish value to an array of labels. */
export function asNamespace(value: NamespaceLike): string[] {
  if (typeof value === "string") return [value];
  return value.map((label) => String(label));
}

/**
 * The names a placeholder is looked up under, in order.
 *
 * LangGraph's own protocol keys are snake_case (`thread_id`, `checkpoint_id`) while a TypeScript
 * caller naturally writes `userId`, and the docs' own examples use `user_id`. Rather than making
 * that a coin flip, `{userId}` also matches a `user_id` entry and vice versa.
 */
function candidateKeys(name: string): string[] {
  const snake = name.replace(/([a-z0-9])([A-Z])/g, "$1_$2").toLowerCase();
  const camel = name.replace(/_([a-z0-9])/g, (_m, c: string) => c.toUpperCase());
  const seen = new Set<string>([name, snake, camel]);
  return [...seen];
}

/**
 * Expand `{placeholder}` labels in `template` from a runnable config.
 *
 * @param template A namespace whose labels may be placeholders, e.g. `["memories", "{userId}"]`.
 * @param config The config to read `configurable` from. When omitted, the ambient config is
 *   pulled via `getConfig()` — which only returns something inside a running graph node or tool.
 * @returns The namespace with every placeholder substituted.
 * @throws `Error` if a placeholder has no value, or if no config is available at all.
 */
export function resolveNamespace(
  template: NamespaceLike = DEFAULT_NAMESPACE,
  config?: { configurable?: Record<string, unknown> },
): string[] {
  const labels = asNamespace(template);
  const placeholders: Array<[number, string]> = [];
  labels.forEach((label, index) => {
    const match = PLACEHOLDER.exec(label);
    if (match) placeholders.push([index, match[1]!]);
  });
  if (placeholders.length === 0) return labels;

  const resolvedConfig = config ?? ambientConfig();
  if (resolvedConfig === undefined) {
    const wanted = placeholders.map(([, name]) => name).join(", ");
    throw new Error(
      `Namespace [${labels.join(", ")}] needs {${wanted}} from the runnable config, but no ` +
        "config was passed and none is available. Either invoke this inside a LangGraph node " +
        'with config={ configurable: { ... } }, or pass a fully-resolved namespace.',
    );
  }

  const configurable = resolvedConfig.configurable ?? {};
  const resolved = [...labels];
  const missing: string[] = [];
  for (const [index, name] of placeholders) {
    const key = candidateKeys(name).find((candidate) => {
      const value = configurable[candidate];
      return value !== undefined && value !== null && value !== "";
    });
    if (key === undefined) missing.push(name);
    else resolved[index] = String(configurable[key]);
  }
  if (missing.length > 0) {
    throw new Error(
      `Namespace [${labels.join(", ")}] could not be resolved: missing ` +
        `${missing.map((name) => `"${name}"`).join(", ")} in config.configurable. ` +
        `Invoke with config={ configurable: { ${missing[0]}: "..." } }.`,
    );
  }
  return resolved;
}

/**
 * Build a `namespace -> userId` resolver that reads one label of the namespace.
 *
 * `namespaceUserId(-1)` pairs with {@link DEFAULT_NAMESPACE}: given `["memories", "alice"]` it
 * yields `"alice"`, so one {@link CortadelStore} can serve every user of a multi-tenant graph
 * while still honouring Cortadel's one-client-per-user rule.
 *
 * @param index Position of the label holding the user id. Negative indices count from the end.
 */
export function namespaceUserId(index: number = -1): (namespace: string[]) => string {
  return function namespaceUserIdResolver(namespace: string[]): string {
    const position = index < 0 ? namespace.length + index : index;
    const label = namespace[position];
    if (label === undefined) {
      throw new Error(
        `Namespace [${namespace.join(", ")}] has no label at index ${index}, so no Cortadel ` +
          "user id could be derived from it.",
      );
    }
    if (label === "") {
      throw new Error(
        `Namespace [${namespace.join(", ")}] has an empty label at index ${index}; a Cortadel ` +
          "user id cannot be blank.",
      );
    }
    return label;
  };
}

File: integrations/langgraph/cortadel_langgraph/_namespace.py
"""Namespace templating and user-id resolution.

LangGraph addresses long-term memory by *namespace* — a tuple of strings such as
``("memories", "alice")`` (see ``langgraph/store/base/__init__.py``, ``BaseStore.put``). Cortadel
addresses it by *user id*, fixed at client construction: "All calls are scoped to the userId given
at construction." Those two models meet here.

Two small utilities:

* :func:`resolve_namespace` expands ``{placeholder}`` labels from a ``RunnableConfig``'s
  ``configurable`` dict, so a single tool or node can serve every user — the same convention
  ``langmem.utils.NamespaceTemplate`` uses.
* :func:`namespace_user_id` builds the ``namespace -> user_id`` callable that
  :class:`~cortadel_langgraph.store.CortadelStore` uses to decide *which* Cortadel client a given
  namespace belongs to.
"""
from __future__ import annotations

import re
from typing import Any, Callable, Mapping, Optional, Sequence, Union

from langgraph.config import get_config

__all__ = [
    "DEFAULT_NAMESPACE",
    "NamespaceLike",
    "UserIdResolver",
    "as_namespace",
    "namespace_user_id",
    "resolve_namespace",
]

DEFAULT_NAMESPACE = ("memories", "{user_id}")
"""The namespace this package defaults to.

Mirrors the shape LangGraph's own long-term-memory docs and ``langmem`` use, and pairs with
``namespace_user_id(-1)`` so the trailing label *is* the Cortadel user id.
"""

NamespaceLike = Union[str, Sequence[str]]
"""A namespace, or a template for one. A bare ``str`` is treated as a one-label namespace."""

UserIdResolver = Union[str, Callable[[tuple[str, ...]], str]]
"""Either a fixed Cortadel user id, or a callable mapping a namespace to one."""

# Only a *whole* label may be a placeholder (``"{user_id}"``, never ``"user-{id}"``). Same
# restriction langmem imposes in ``NamespaceTemplate._get_key``; keeping it means a namespace label
# can never be half-literal, which would make the reverse mapping (namespace -> user id) ambiguous.
_PLACEHOLDER = re.compile(r"^\{(\w+)\}$")


def as_namespace(value: NamespaceLike) -> tuple[str, ...]:
    """Normalise a namespace-ish value to a tuple of labels."""
    if isinstance(value, str):
        return (value,)
    return tuple(str(label) for label in value)


def resolve_namespace(
    template: NamespaceLike,
    config: Optional[Mapping[str, Any]] = None,
) -> tuple[str, ...]:
    """Expand ``{placeholder}`` labels in ``template`` from a ``RunnableConfig``.

    Args:
        template: A namespace whose labels may be placeholders, e.g. ``("memories", "{user_id}")``.
        config: The runnable config to read ``configurable`` from. When ``None``, the ambient
            config is pulled via ``langgraph.config.get_config()`` — which only works inside a
            running graph node, tool, or task.

    Returns:
        The namespace with every placeholder substituted.

    Raises:
        ValueError: If a placeholder has no value, or if the ambient config is unavailable.
    """
    labels = as_namespace(template)
    placeholders = [
        (index, match.group(1))
        for index, label in enumerate(labels)
        if (match := _PLACEHOLDER.match(label))
    ]
    if not placeholders:
        return labels

    if config is None:
        try:
            config = get_config()
        except RuntimeError as exc:  # raised verbatim by langgraph.config.get_config
            wanted = ", ".join(name for _, name in placeholders)
            raise ValueError(
                f"Namespace {labels} needs {{{wanted}}} from the runnable config, but no config "
                "was passed and none is available. Either invoke this inside a LangGraph node "
                'with config={"configurable": {...}}, or pass a fully-resolved namespace.'
            ) from exc

    configurable = config.get("configurable") or {} if config else {}
    resolved = list(labels)
    missing: list[str] = []
    for index, name in placeholders:
        value = configurable.get(name)
        if value is None or value == "":
            missing.append(name)
        else:
            resolved[index] = str(value)
    if missing:
        raise ValueError(
            f"Namespace {labels} could not be resolved: "
            f"missing {', '.join(repr(name) for name in missing)} in config['configurable']. "
            f"Invoke with config={{'configurable': {{{missing[0]!r}: '...'}}}}."
        )
    return tuple(resolved)


def namespace_user_id(index: int = -1) -> Callable[[tuple[str, ...]], str]:
    """Build a ``namespace -> user_id`` resolver that reads one label of the namespace.

    ``namespace_user_id(-1)`` pairs with :data:`DEFAULT_NAMESPACE`: given ``("memories", "alice")``
    it yields ``"alice"``, so one :class:`~cortadel_langgraph.store.CortadelStore` can serve every
    user of a multi-tenant graph while still honouring Cortadel's one-client-per-user rule.

    Args:
        index: Position of the label holding the user id. Negative indices count from the end.
    """

    def _resolve(namespace: tuple[str, ...]) -> str:
        try:
            label = namespace[index]
        except IndexError as exc:
            raise ValueError(
                f"Namespace {namespace} has no label at index {index}, so no Cortadel user id "
                "could be derived from it."
            ) from exc
        if not label:
            raise ValueError(
                f"Namespace {namespace} has an empty label at index {index}; a Cortadel user id "
                "cannot be blank."
            )
        return label

    _resolve.__name__ = f"namespace_user_id({index})"
    return _resolve
